conda/notices/core.py

```diff
@@ -35,69 +36,111 @@ def display_notices(
 
-    views.print_more_notices_message(num_total_notices, len(channel_notices), num_viewed_notices)
+    views.print_more_notices_message(
+        channel_notice_set.total_number_channel_notices,
+        len(channel_notice_set.channel_notices),
+        channel_notice_set.viewed_channel_notices,
+    )
 
 
 def notices(func):
     """
     Wrapper for "execute" entry points for subcommands.
 
+    If channel notices need to be fetched, we do that first and then
+    run the command normally. We then display these notices at the very
+    end of the command output so that the user is more likely to see them.
+
+    This ordering was specifically done to address the following bug report:
+        - https://github.com/conda/conda/issues/11847
+
     Args:
         func: Function to be decorated
     """
-
     @wraps(func)
     def wrapper(*args, **kwargs):
-        return_value = func(*args, **kwargs)
-
-        if is_channel_notices_enabled(context):
-            display_notices(
+        if is_channel_notices_enabled(context) and is_channel_notices_cache_expired():
+            channel_notice_set = retrieve_notices(
                 limit=context.number_channel_notices,
                 always_show_viewed=False,
                 silent=True,
             )
+            return_value = func(*args, **kwargs)
+            display_notices(channel_notice_set)
+
+            return return_value
 
-        return return_value
+        else:
+            return func(*args, **kwargs)
 
     return wrapper
 
 
 def get_channel_name_and_urls(
-    channels: [Sequence[Channel]],
-) -> Sequence[Tuple[ChannelUrl, ChannelName]]:
-    """Return a sequence of Channel URL and name"""
+    channels: Sequence[Union[Channel, MultiChannel]],
+) -> Sequence[tuple[ChannelUrl, ChannelName]]:
+    """
+    Return a sequence of Channel URL and name tuples.
+
+    This function handles both Channel and MultiChannel object types.
+    """
 
     def ensure_endswith(value: str, ends: str) -> str:
         return value if value.endswith(ends) else f"{value}{ends}"
 
     def join_url(value: str, join_val: str) -> str:
-        return parse.urljoin(ensure_endswith(value, "/"), join_val)
+        """Using this method because urllib.parse.urljoin does not support s3:// URLs"""
+        value = ensure_endswith(value, "/")
+        return f"{value}{join_val}"
 
-    return tuple(
-        (join_url(base_url, NOTICES_FN), channel.name or channel.location)
-        for channel in channels
-        for base_url in channel.base_urls
-    )
+    channel_name_and_urls = []
+
+    for channel in channels:
+        name = channel.name or channel.location
+
+        if type(channel) is Channel:
+            channel_name_and_urls.append((join_url(channel.base_url, NOTICES_FN), name))
+
+        elif type(channel) is MultiChannel:
+            for url in channel.base_urls:
+                channel_name_and_urls.append((join_url(url, NOTICES_FN), name))
```

Comment:
I thought `isinstance` checks are still sensible?

```suggestion
        if isinstance(channel, Channel):
            channel_name_and_urls.append((join_url(channel.base_url, NOTICES_FN), name))

        elif isinstance(channel, MultiChannel):
            for url in channel.base_urls:
                channel_name_and_urls.append((join_url(url, NOTICES_FN), name))
```

Reply:
That's what I thought at first too, but with the changes you proposed, you will always go to the first block because `MultiChannel` is a sub-class of `Channel`. That's why I put the more strict "type" checking in there.

I think I have a better idea for how to do this though now that I'm seeing this code again:
```suggestion
        for url in channel.base_urls:
            channel_name_and_urls.append((join_url(url, NOTICES_FN), name))
```

Both classes define their own `base_urls` property that returns a tuple.

Reply:
Ah in that case you can also change the order of the if statements to first check for `MultiChannel` and then `Channel`